controlpanel/api/rules.py

```diff
@@ -0,0 +1,194 @@
+"""
+Rules and permissions (and object permissions) using django-rules
+"""
+
+from rules import add_perm, is_active, is_authenticated, is_superuser, predicate
+
+from controlpanel.api.models import (
+    App,
+    AppS3Bucket,
+    Parameter,
+    S3Bucket,
+    UserApp,
+    UserS3Bucket,
+)
+from controlpanel.api.tools import ToolDeployment
+
+
+@predicate
+def is_app_admin(user, obj):
+    """
+    Check whether `user` is an admin for either;
+    - the `obj` `App`
+    - the `app` of the `obj` `AppS3Bucket`
+    - the `app` of the `obj` `UserApp`
+
+    :param user User: The user to check
+    :param obj App|AppS3Bucket|UserApp: The object to check
+    """
+    if obj is None:
+        return True
+
+    if isinstance(obj, App):
+        return user in obj.admins
+
+    if isinstance(obj, AppS3Bucket):
+        return is_app_admin(user, obj.app)
+
+    if isinstance(obj, UserApp):
+        return obj.is_admin
+
+    # XXX raise exception?
+    return False
+
+
+add_perm('api.list_app', is_authenticated)
+add_perm('api.create_app', is_authenticated & is_superuser)
+add_perm('api.retrieve_app', is_authenticated & is_app_admin)
+add_perm('api.update_app', is_authenticated & is_superuser)
+add_perm('api.destroy_app', is_authenticated & is_superuser)
+add_perm('api.add_app_customer', is_authenticated & is_app_admin)
+add_perm('api.remove_app_customer', is_authenticated & is_app_admin)
+add_perm('api.add_app_admin', is_authenticated & is_superuser)
+add_perm('api.remove_app_admin', is_authenticated & is_superuser)
+add_perm('api.add_app_bucket', is_authenticated & is_superuser)
+add_perm('api.remove_app_bucket', is_authenticated & is_superuser)
+add_perm('api.view_app_logs', is_authenticated & is_app_admin)
+
+
+@predicate
+def has_bucket_access(user, obj, **kwargs):
+    """
+    Check whether `user` has access to;
+    - the `obj` `S3Bucket`
+    - the `s3bucket` of the `obj` `AppS3Bucket`
+    - the `s3bucket` of the `obj` `UserS3Bucket`
+
+    :param user User: The user to check
+    :param obj AppS3Bucket|S3Bucket|UserS3Bucket: The object to check
+    :param kwargs dict: Additional search criteria
+    """
+    if obj is None:
+        return True
+
+    if isinstance(obj, S3Bucket):
+        try:
```

Comment:
You could change to not use the exceptions to something like:

```python
# not tested/tried!
return UserS3Bucket.objects.filter(user=user, s3bucket=obj, **kwargs).exists()
```

Assuming here we just want to check if there is at least a record for the user.

(same results but without the exceptions which probably are a bit drastic in this case)

Reply:
Ah! I knew there must be a simpler way 👍